### PR TITLE
[webui][api] Change SendEventEmails to 1 event per job.

### PR DESCRIPTION
**Pull request by evanrolfe**
To move the tracking of jobs (if they have been performed yet, or if they have been failed) from the events table to the delayed_jobs gem since that what is already designed for.

### CODE REVIEW

File: src/api/app/jobs/send_event_emails_job.rb [hunk 1]
```diff
@@ -1,25 +1,23 @@
...
+    begin
+      create_rss_notifications(event)
+      EventMailer.event(subscribers, event).deliver_now
+    rescue StandardError => e
+      Airbrake.notify(e, event_id: event.id)
+    ensure
+      event.update_attributes(mails_sent: true)
     end
-    true
   end
```
Comment (bgeuken) [inline src/api/app/jobs/send_event_emails_job.rb]: @evanrolfe Does perform always set the mails_sent attribute to true?

**Comment by coolo** [comment]
It should be noted, that we moved *away* from one job per email because delayed job sucks heavily if you have more than 10T jobs. The overhead of grabbing a job becomes so painful that there was no way to recover from an event backlog. If delayedjob was down over the weekend (you do have panic notifications from monitoring meanwhile, right?) you have easily 400T events in the backend that await processing. 

So basically my question is: what changed in the problem space since https://github.com/openSUSE/open-build-service/pull/1324 ?

**Comment by evanrolfe** [comment]
> So basically my question is: what changed in the problem space since #1324 ?

@coolo I had not seen that PR.. Moises, Henne and me decided to move to 1 event per job after we ran into many problems with the events table (mysql deadlock errors etc.) when trying to implement the daily digest email feature. A similar thing happened with the events table, one job for an event failed and then the queue started growing to millions of events in the queue. So we did [a review of the system](https://github.com/openSUSE/open-build-service/wiki/Event-Subsystem-Architecture-Review) and decided on 1 event per job.

> It should be noted, that we moved away from one job per email because delayed job sucks heavily if you have more than 10T jobs.

Right ok, I'll investigate that then, the PR you posted is from 2015 so its possible delayed job has improved but I will make sure we can handle this. If delayed job still can handle that then I would suggest finding an alternative to delayed_job which can. Also, by 10T do you mean 10,000? I'm not familiar with that syntax..

> If delayedjob was down over the weekend (you do have panic notifications from monitoring meanwhile, right?) you have easily 400T events

This happened on the friday after christmas that delayed_job had been down for some time and there was a backlog of jobs but delayed_job processed them fine. FYI [CreateProjectLogEntryJob](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/jobs/create_project_log_entry_job.rb) formerly known as ProjectLogRotate has been on 1 event / 1 job since before christmas and it handled it fine. Though its true this is much less intensive then SendEventEmails.. And I'm not sure about panic notifications.

**Comment by coolo** [comment]
And what was the peak number of jobs?

**Comment by evanrolfe** [comment]
I can't remember exactly but it was in the 10,000s range...

**Comment by coolo** [comment]
See https://github.com/collectiveidea/delayed_job/issues/581

**Comment by hennevogel** [comment]
> So basically my question is: what changed in the problem space since #1324 ?

[A lot of things have changed](https://github.com/openSUSE/open-build-service/wiki/Event-Subsystem-Architecture-Review). The number of jobs has gone down (no backend plugins anymore), we are using many different queues with the right **[tm]** concurrency, we got rid of most of the state tracking in the Events table, ActiveJob etc. etc. And it's also not like the Event's table helped when something crashed over the weekend...

**Comment by evanrolfe** [comment]
And now for some data.. By running this [script](https://gist.github.com/evanrolfe/31cef214adf286f9c994ab07a45049f6) I created ~500k jobs in the delayed_jobs table (they were all for CreateProjectLogEntryJob). Then running this command to start 10 DJ workers to work off that queue:
```
script/delayed_job.api.rb start --queue=project_log_rotate -n 10
```

And then with this function:
```ruby
def calculate_jobs_rate
  jobs_num_start = Delayed::Job.count
  sleep 60
  jobs_num_end = Delayed::Job.count

  diff = jobs_num_start - jobs_num_end

  puts "#{diff} jobs completed in 60 sec"
end
```

We run it three times like the good scientists we are:
```
irb(main):011:0> Delayed::Job.count
=> 522515
irb(main):012:0> calculate_jobs_rate
=> 47 jobs completed in 60 sec
irb(main):013:0> calculate_jobs_rate
=> 46 jobs completed in 60 sec
irb(main):014:0> calculate_jobs_rate
=> 46 jobs completed in 60 sec
```

And then doing the same thing but this time with ~2000 jobs we get:
```
irb(main):037:0> Delayed::Job.count
=> 2037
irb(main):039:0> calculate_jobs_rate
=> 1615 jobs completed in 60 sec
```

So at 500k jobs  we process them at 46 jobs / min. At this rate it would take nearly 8 days to process them, though the real figure while be significantly less since that calculation does not consider that the rate would be increasing as the number of jobs decreases. Its not an ideal figure but also not fatal either. Given the options I still think we should move to 1 event / 1 job.

**Comment by coolo** [comment]
remember that another 600K jobs pile up in these 8 days.

**Comment by coolo** [comment]
(this is not a theory, we've been there)

**Comment by evanrolfe** [comment]
> (this is not a theory, we've been there)

Then maybe the solution is to improve monitoring of delayed_jobs so we get alerted the second it goes down?

**Comment by hennevogel** [comment]
> Then maybe the solution is to improve monitoring of delayed_jobs so we get alerted the second it goes down?

And to mitigate this by some other means, like more workers for this queue or something. But all of this discussion is moot, as the `Event` table does not help. Exactly the same happens if we have too many `Events` now.

**Comment by coolo** [comment]
the problem with the delayed_job table is a bit special - as you can read in the referenced delayed_job issue. The way they query the next job is just very mysql unfriendly

**Comment by Ana06** [comment]
@coolo 

> the problem with the delayed_job table is a bit special - as you can read in the referenced delayed_job issue. The way they query the next job is just very mysql unfriendly

what about fixing how the delayed_job query the next job? 🤔

**Comment by evanrolfe** [comment]
>  Exactly the same happens if we have too many Events now.

I don't think this is entirely true? In the past we've had > 2 million events in the table and they were processed within a day (even with new events being added as they were processed). If there were 2 million delayed_jobs in the table then this would come to a grinding halt. That said there are other problems with the Events table like mysql deadlocks and other problems where one failing event can prevent the other events from processing.

I've opened PR https://github.com/openSUSE/open-build-service/pull/4370 which sends an email if the number of delayed jobs in reaches a limit.

IMO the best long-term solution is to use something other than delayed_job which can handle 500k jobs in the queue. Though I would first monitor how delayed_jobs performs once we move everything to 1 event / job and see if it is really needed or not.

> what about fixing how the delayed_job query the next job? :thinking:

We can put a lot of time and effort into trying to improve delayed_job so it can scale better or we can just switch to one of the many async job gems that does that already..

**Comment by hennevogel** [comment]
> I don't think this is entirely true? In the past we've had > 2 million events in the table and they were processed within a day (even with new events being added as they were processed).

I have a different memory about this. I know I personally have manually deleted all `Event`s that wouldn't be processed by email or rss (like `BuildSuccess` or other high traffic `Event`s) in those cases.

**Comment by evanrolfe** [comment]
> I have a different memory about this.

Looks like will have to agree to disagree then :)

**Comment by coolo** [comment]
I too deleted Events, but not because the processing rate of DJs dropped, but because a SendEmailEvent takes Xms per Event, even if no email is sent. And if you have 2M of them, these X sum up. And there are a lot of events where we know that they won't send emails - so removing them helps the DJs.

This especially bad when the source_watcher implementation slowed down every job to a minute.

**Comment by hennevogel** [comment]
Codecov hangs, merging.

**Comment by coolo** [comment]
So https://github.com/openSUSE/open-build-service/issues/4473 tells me this is deployed now?

**Comment by coolo** [comment]
Where can I see your monitor check about the current job count? At the moment it grows about 300 per minute

**Comment by evanrolfe** [comment]
in production server:
```
run_in_api rails runner script/delayed_job_stats.rb
```
